sql/parser2/eval.go

```diff
@@ -284,174 +536,46 @@ func evalComparisonExpr(expr *ComparisonExpr, env Env) (sqlwire.Datum, error) {
 		return null, err
 	}
 
-	op := expr.Operator
-
 	if left.IsNull() || right.IsNull() {
 		return null, nil
 	}
 
-	var typ opType
-	var v bool
-
+	not := false
+	op := expr.Operator
 	switch op {
-	case EQ:
-		typ, left, right, err = prepareComparisonArgs(left, right)
-		switch typ {
-		case intOp:
-			v = *left.IntVal == *right.IntVal
-		case uintOp:
-			v = *left.UintVal == *right.UintVal
-		case floatOp:
-			v = *left.FloatVal == *right.FloatVal
-		case stringOp:
-			v = *left.StringVal == *right.StringVal
-		default:
-			panic(fmt.Sprintf("unsupported op type: %d", typ))
-		}
-
-	case LT:
-		typ, left, right, err = prepareComparisonArgs(left, right)
-		switch typ {
-		case intOp:
-			v = *left.IntVal < *right.IntVal
-		case uintOp:
-			v = *left.UintVal < *right.UintVal
-		case floatOp:
-			v = *left.FloatVal < *right.FloatVal
-		case stringOp:
-			v = *left.StringVal < *right.StringVal
-		default:
-			panic(fmt.Sprintf("unsupported op type: %d", typ))
-		}
-
-	case LE:
-		typ, left, right, err = prepareComparisonArgs(left, right)
-		switch typ {
-		case intOp:
-			v = *left.IntVal <= *right.IntVal
-		case uintOp:
-			v = *left.UintVal <= *right.UintVal
-		case floatOp:
-			v = *left.FloatVal <= *right.FloatVal
-		case stringOp:
-			v = *left.StringVal <= *right.StringVal
-		default:
-			panic(fmt.Sprintf("unsupported op type: %d", typ))
-		}
-
+	case NE:
+		// NE(left, right) is implemented as !EQ(left, right).
+		not = true
+		op = EQ
 	case GT:
-		typ, left, right, err = prepareComparisonArgs(left, right)
-		switch typ {
-		case intOp:
-			v = *left.IntVal > *right.IntVal
-		case uintOp:
-			v = *left.UintVal > *right.UintVal
-		case floatOp:
-			v = *left.FloatVal > *right.FloatVal
-		case stringOp:
-			v = *left.StringVal > *right.StringVal
-		default:
-			panic(fmt.Sprintf("unsupported op type: %d", typ))
-		}
-
+		// GT(left, right) is implemented as LT(right, left)
+		op = LT
+		left, right = right, left
 	case GE:
-		typ, left, right, err = prepareComparisonArgs(left, right)
-		switch typ {
-		case intOp:
-			v = *left.IntVal >= *right.IntVal
-		case uintOp:
-			v = *left.UintVal >= *right.UintVal
-		case floatOp:
-			v = *left.FloatVal >= *right.FloatVal
-		case stringOp:
-			v = *left.StringVal >= *right.StringVal
-		default:
-			panic(fmt.Sprintf("unsupported op type: %d", typ))
-		}
+		// GE(left, right) is implemented as LE(right, left)
+		op = LE
+		left, right = right, left
+	}
 
-	case NE:
-		typ, left, right, err = prepareComparisonArgs(left, right)
-		switch typ {
-		case intOp:
-			v = *left.IntVal != *right.IntVal
-		case uintOp:
-			v = *left.UintVal != *right.UintVal
-		case floatOp:
-			v = *left.FloatVal != *right.FloatVal
-		case stringOp:
-			v = *left.StringVal != *right.StringVal
-		default:
-			panic(fmt.Sprintf("unsupported op type: %d", typ))
+	f := cmpOps[cmpArgs{op, reflect.TypeOf(left), reflect.TypeOf(right)}]
+	if f != nil {
```

Comment:
Yep, if the function cannot be found the desired operation does not exist for the argument types. For example, you can't compare an int with a string.